Refactor UI for better UX & DX

599316527 opened the pull request:
## Preview

Compiled result is here. And some screenshots. 

https://github.com/599316527/teslausb/releases/tag/new_ui_preview_0

## Related tech

* Vue: A data-driven web framework. https://vuejs.org/
* Veui: A ui component library. https://veui.dev/en-US

## TODO

- [x] Shortcuts for viewers: keyboard & (double)click on video
- [x] Performance improvment for player
- [x] Solve cases to keep videos in synchronous

## How to dev

1. Installl [Nodejs](https://nodejs.org/).
2. Into ui folder, run `npm install` to install dependencies.
3. Set env variables:
      * PROXT_TARGET=http://xxx.xxx.xxx.xxx  PI's web page url which is used inside dev server to load resource from PI
      * PORT=xxxx  dev server's port 
4. Run `npm run serve` to start dev server.

599316527 committed:
First commit of UI

599316527 committed:
Responsive header links

599316527 committed:
+ LogView

599316527 committed:
+ i18n

599316527 committed:
+ Viewer Layout

599316527 committed:
Viewer: episode select

599316527 committed:
EpisodeSelect tooltip for thumb

599316527 committed:
video player

599316527 committed:
Improve multi-tracks video player

Commit:
enable map for viewer

marcone commented:
This is very nice, but in its current form I can't accept this PR because as uploaded here it's just a UI definition and wouldn't actually install on a device during setup. I guess that's why you said it is a preview.

Functionality wise, there are some things I like, and some I don't:
- archiveloop and setup logs don't scroll to the bottom and automatically pick up new changes to the logs
- video viewer doesn't indicate the time of a Sentry event on the timeline (the triangle below the progress bar in the current web ui, similar to the red dot in the car's player).
- I like the multiple playback speeds
- the scrolling video selector gets a bit unwieldy and hard to read when there is a large number of Sentry events (stronger indentation, different font styles, or a deeper-level menu might help here?)
- I like that it splits daily recordings into segments
- I like that it shows a preview image when hovering over events in the selector
- I like the way the date/time is shown
- I like that it shows the name of the location of the Sentry event
- bug: navigating to the parent directory in the recordings tab goes back to the old UI :)

Some logistical concerns I have:
- I don't actually know Vue or Node.js, so continuing development with this would be a major learning effort for me
- I don't speak Chinese, so even if I did continue development in Vue I wouldn't be able to update the existing translations for new/changed functionality

I would love for people to be able to use this new UI though, so perhaps some kind of runtime UI switcher could be added?
Could be as simple as a button on each UI that switches to the other. (the preview demo is self-contained in a "dist" folder, so switching is easily done)

599316527 commented:
Thanks for your detailed relay.

> but in its current form I can't accept this PR

It's my fault. I should create this PR as draft since there are still a lot work to complete. I can't convert it to draft after creating because I don't have write permission to this repo. You can mark it as draft. I'll  push more fixes later. Once I'm done all improvement, I will let you know.

> I don't actually know Vue or Node.js, so continuing development with this would be a major learning effort for me

I think you must put much effort on building a web app to display on-going logs or playing recordings without latency... But writing vanilla js / css to build so a complex single page app isn't a good idea so I proposal we shall using mordern UI framework and related component library to simplify development and acquire better user experience. And I'd love to continue the following jobs of frontend part. 

> it's just a UI definition and wouldn't actually install on a device during setup

WIth mordern web toolchain, compilation is required to transform, merge and compress js/css into html entry and minified js/css files which are delivered to end-users' browser finally. 

You mentioned `dist` folder above. Yes, it's the folder created by `npm run build` and contains compiled files. With setting `publicPath` field which is defined in [`vue.config.js`](https://cli.vuejs.org/config/#publicpath) to empty string `""` would make referances to js/css files relative to entry html so that the entire `dist` folder can be just put in nginx's document folder as sub-folder for public access by suffixing sub-folder's name to url, such as `http://192.168.1.100/dist`.

So we can rename the folder from `dist` to `v2` or `new` and pack it into `UI.tgz`. When installing, downloading and extracting it into `var/www/html` are the only steps. The url is `/v2` or `/new` depends on what's name the folder is. Links to `/` and  '/v2` can be added for old/new UI switching.

> the scrolling video selector gets a bit unwieldy

I designed a fullscreen selector for mobile. You can see it on phone or narrow your browser's width to less than 500px. I think this selector could be also for PC as dialog for larger viewport of candidates.

You may not found out the selector is searchable. When you click the selector, an input box is focus. With this status, you can type numbers to filter videos. The feature might be take to dialog as well.

> I don't speak Chinese, so even if I did continue development in Vue I wouldn't be able to update the existing translations for new/changed functionality

There are some strategies to solve it.

* If there's no translation for the term, fallback to english. English term always exists.
    * Shortage: english terms on non-english UI until volunteers supply lacked terms for the language
* Once there is any add or change to terms, translations are recognized as out-of-date and will not be shipped with bundle until volunteers supply lacked terms for the language and mark it up-to-date.
    * Shortage: any changes to terms will discard all translation

I prefer the fist one as normal strategy. And if the translation is behind to much, we discard it.

599316527 committed:
Merge branch 'marcone:main-dev' into ui_dev

599316527 committed:
Improve Episode Selector

* Dialog picker for PC as well
* Date range filter in dialog picker
* Dropdown shows date/time in different columns

599316527 committed:
[Viewer] Improve control for playing clips synchronously and fluently

599316527 committed:
Links to switch between new and legacy versions

599316527 committed:
workflow for building & release of UI

599316527 committed:
[Viewer] show video error message

599316527 committed:
bug fix & locale fallback

599316527 commented:
Hi, @marcone . I think it's ready now.

Latest UI tarball can be found [here](https://github.com/599316527/teslausb/releases/tag/v3.1).

Fixes:

* Log will scroll to bottom on load
* Time of Sentry event is marked in progress bar
* Video selector enables dialog view which can be triggered by clicking left part of selector for PC as well
* Day and time list are separated in video dropdown selector
* Link to parent directory is hidden in root page
* Rewrite video controller for better synchronization
* Keyboard shortcuts for video player: <kbd>Space</kbd>, <kbd>←</kbd>, <kbd>→</kbd>

Adds:

* Links for switching between old and new version
* Translation fallback if terms are not exist in current locale

Distribution:

1. Packing and publish UI tarball by Github Actions
    * Run workflow manually and UI tarball will be uploaded to latest release
2. Download and extract UI tarball to `/var/www/html` in `configure-web`
    * UI tarball is from the latest release (uploaded in previous step)

599316527 committed:
fix compatible for mobile

marcone commented on .github/workflows/build-ui.yml:
```diff
@@ -0,0 +1,45 @@
+        npm ci
+        ./build.sh
+
+    - name: Get latest release of NodeJS
```
This section doesn't seem to have anything to do with NodeJS. If I understand this correctly, it would get the latest version of the repository in which it is run, in this case teslausb.

599316527 replied:
Yes, the name should be "Get latest release of current repo".

marcone commented on .github/workflows/build-ui.yml:
```diff
@@ -0,0 +1,45 @@
+        repository: ${{ github.repository }}
+        type: "latest"
+
+    - name: Upload files to a GitHub release
```
Is it possible to check the generated teslausb-ui.tgz into the repo, instead of making it a new release? I think it will be confusing if https://github.com/marcone/teslausb/releases becomes a mix of UIs and OS images.

599316527 replied:
It's [possible](https://github.community/t/push-built-files-to-repo-using-actions/18186) to push generated files to current repo but I don't think it's a good idea. Repo is not designed to store binary files which cause slow checkout, occupy large amounts of repo size (since all pushed files are in git's history) and may exceeds github's maxiumn limitation sooner or later. 

To clarify releases of OS Images and UIs, publishing them into different releases could be more reasonable? Or UIs can be published to npm's registry which is owned by Github as well.

marcone reviewed (commented):
When I tried this locally, I got the following error when building:
```
> TeslaUSB@0.1.0 build /home/marcone/tesla/teslausb/ui
> vue-cli-service build

 ERROR  Error loading /home/marcone/tesla/teslausb/ui/vue.config.js:
 ERROR  Error: Cannot find module 'unplugin-vue-components/webpack'
Error: Cannot find module 'unplugin-vue-components/webpack'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/home/marcone/tesla/teslausb/ui/vue.config.js:46:13)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! TeslaUSB@0.1.0 build: `vue-cli-service build`
npm ERR! Exit status 1
```
I _think_ I pretty much followed the steps you described in the readme. I ran `sudo apt install nodejs npm`, then `npm install` in the ui folder, then I ran the `build.sh` script. Did I miss a step somewhere?
This is on Ubuntu 20.04 LTS.

599316527 commented on setup/pi/configure-web.sh:
```diff
@@ -32,6 +32,11 @@ ln -sf /etc/nginx/sites-available/teslausb.nginx /etc/nginx/sites-enabled/defaul
 g++ -o /root/cttseraser -D_FILE_OFFSET_BITS=64 "$SOURCE_DIR/teslausb-www/cttseraser.cpp" -lstdc++ -lfuse
 
+# install new UI (compiled js/css files)
+wget -O /tmp/teslausb-ui.tgz https://github.com/"$REPO"/teslausb/releases/latest/download/teslausb-ui.tgz
```
If UIs are published to npm's registry, the url here shall be changed to one of:

* Getting latest tarball url from npm's API ([ref](https://stackoverflow.com/a/51318312/1872976))
    ```
    curl https://registry.npmjs.org/PACKAGE-NAME/ \
         | jq '.versions[."dist-tags".latest].dist.tarball'
    ```
* Using [unpkg.com](https://unpkg.com/) to resolve latest version automatically

marcone replied:
Note that this line currently does not work, because the REPO variable isn't normally exported. It would need to be set and exported explicitly from the setup .conf file.

I prefer to check the generated html/css/js files in to the repo (or a .tgz file of those) and download using the `get_script` function. This would also support having different development branches.

599316527 commented:
According to the error `Cannot find module 'unplugin-vue-components/webpack'`, the `webpack` file is not exist. 

First, run `ls node_modules/unplugin-vue-components` to check whether the package is installed properly. (Is there any eror occurred during `npm install`?). Or try [`npm ci`](https://docs.npmjs.com/cli/v8/commands/npm-ci) to make a clean install. 

I found that  the `unplugin-vue-components` package [exports](https://github.com/antfu/unplugin-vue-components/blob/c81257e2ea7c55686223bb8e9f18316602f2542c/package.json?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L44-L46) it's sub-files by `exports` field in package.json which is introduced in node v12.7.0. So can you run `node -v` and `npm -v` to print node/npm's version? Check whether your node supports the `exports` field.

In my opinion, downloading tarball from [node.js](https://nodejs.org/en/download/) and add `bin` folder (npm is also included in this folder) to `PATH` env is more reliable than installing by package maneger.

marcone commented:
> I found that the `unplugin-vue-components` package [exports](https://github.com/antfu/unplugin-vue-components/blob/c81257e2ea7c55686223bb8e9f18316602f2542c/package.json?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L44-L46) it's sub-files by `exports` field in package.json which is introduced in node v12.7.0.

That's the problem then. My version is v10.19.0

marcone commented:
> In my opinion, downloading tarball from [node.js](https://nodejs.org/en/download/) and add `bin` folder (npm is also included in this folder) to `PATH` env is more reliable than installing by package maneger.

I did, and it's still not working:
```
~/tesla/teslausb/ui$ node -v
v16.13.2
~/tesla/teslausb/ui$ npm -v
8.3.2
~/tesla/teslausb/ui$ npm install

up to date, audited 1293 packages in 3s

88 packages are looking for funding
  run `npm fund` for details

70 vulnerabilities (2 low, 55 moderate, 13 high)

To address issues that do not require attention, run:
  npm audit fix

To address all issues (including breaking changes), run:
  npm audit fix --force

Run `npm audit` for details.
~/tesla/teslausb/ui$ ./build.sh
++ dirname ./build.sh
+ BASEDIR=.
+ cd .
+ npm run build

> TeslaUSB@0.1.0 build
> vue-cli-service build


⠙  Building for production...

 ERROR  Failed to compile with 1 error                                                                                                    10:50:45 AM

 error  in ./src/main.js

Syntax Error: Thread Loader (Worker 0)
Cannot find module '@vue/cli-plugin-babel/preset'
Require stack:
- /home/marcone/tesla/teslausb/ui/node_modules/@babel/core/lib/config/files/plugins.js
- /home/marcone/tesla/teslausb/ui/node_modules/@babel/core/lib/config/files/index.js
- /home/marcone/tesla/teslausb/ui/node_modules/@babel/core/lib/index.js
- /home/marcone/tesla/teslausb/ui/node_modules/babel-loader/lib/index.js
- /home/marcone/tesla/teslausb/ui/node_modules/loader-runner/lib/loadLoader.js
- /home/marcone/tesla/teslausb/ui/node_modules/loader-runner/lib/LoaderRunner.js
- /home/marcone/tesla/teslausb/ui/node_modules/thread-loader/dist/worker.js
    at loadPreset.next (<anonymous>)
    at createDescriptor.next (<anonymous>)


 @ multi ./src/main.js

 ERROR  Build failed with errors.

```
In addition, I can't seem to make that warning about vulnerabilities go away. Running `npm audit fix` as instructed just ends up printing the same message about 70 vulnerabilities.

599316527 commented:
It seems weird. Looks like `node_modules/@vue/cli-plugin-babel` does not exist. But since `./build.sh` ran on Github Actions successfully, there's no way that leaking dependencies. 

And I created a clean OS (Ubuntu 20.04.3 LTS) on google cloud but could not reproduce the issue.

<details>
<summary>Records of trying to reproduce</summary>

```
hekai1229@instance-1:~$ ls
hekai1229@instance-1:~$ mkdir local
hekai1229@instance-1:~$ cd local/
hekai1229@instance-1:~/local$ wget https://nodejs.org/dist/v16.13.2/node-v16.13.2-linux-x64.tar.xz
--2022-01-24 02:22:56--  https://nodejs.org/dist/v16.13.2/node-v16.13.2-linux-x64.tar.xz
Resolving nodejs.org (nodejs.org)... 104.20.22.46, 104.20.23.46, 2606:4700:10::6814:172e, ...
Connecting to nodejs.org (nodejs.org)|104.20.22.46|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 21619524 (21M) [application/x-xz]
Saving to: ‘node-v16.13.2-linux-x64.tar.xz’
node-v16.13.2-linux-x64.tar.xz       100%[===================================================================>]  20.62M  --.-KB/s    in 0.07s   
2022-01-24 02:22:56 (284 MB/s) - ‘node-v16.13.2-linux-x64.tar.xz’ saved [21619524/21619524]
hekai1229@instance-1:~/local$ tar -xf node-v16.13.2-linux-x64.tar.xz 
hekai1229@instance-1:~/local$ ls
node-v16.13.2-linux-x64  node-v16.13.2-linux-x64.tar.xz
hekai1229@instance-1:~/local$ export PATH=~/local/node-v16.13.2-linux-x64/bin:$PATH
hekai1229@instance-1:~/local$ node -v
v16.13.2
hekai1229@instance-1:~/local$ npm -v
8.1.2

hekai1229@instance-1:~/local$ git clone --depth=1 https://github.com/599316527/teslausb.git
Cloning into 'teslausb'...
remote: Enumerating objects: 166, done.
remote: Counting objects: 100% (166/166), done.
remote: Compressing objects: 100% (156/156), done.
remote: Total 166 (delta 1), reused 64 (delta 0), pack-reused 0
Receiving objects: 100% (166/166), 372.58 KiB | 3.13 MiB/s, done.
Resolving deltas: 100% (1/1), done.
hekai1229@instance-1:~/local$ cd teslausb/
hekai1229@instance-1:~/local/teslausb$ git branch
* ui_dev
hekai1229@instance-1:~/local/teslausb$ cd ui/
hekai1229@instance-1:~/local/teslausb/ui$ npm install
npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is 
known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated source-map-resolve@0.5.3: See https://github.com/lydell/source-map-resolve#deprecated
npm WARN deprecated source-map-url@0.4.1: See https://github.com/lydell/source-map-url#deprecated
npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
npm WARN deprecated querystring@0.2.0: The querystring API is considered Legacy. new code should use the URLSearchParams API instead.
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated html-webpack-plugin@3.2.0: 3.x is no longer supported
npm WARN deprecated @hapi/topo@3.1.6: This version has been deprecated and is no longer supported or maintained
npm WARN deprecated svgo@1.3.2: This SVGO version is no longer supported. Upgrade to v2.x.x.
npm WARN deprecated @hapi/bourne@1.3.2: This version has been deprecated and is no longer supported or maintained
npm WARN deprecated @hapi/address@2.1.4: Moved to 'npm install @sideway/address'
npm WARN deprecated @hapi/hoek@8.5.1: This version has been deprecated and is no longer supported or maintained
npm WARN deprecated chokidar@2.1.8: Chokidar 2 will break on node v14+. Upgrade to chokidar 3 with 15x less dependencies.
npm WARN deprecated chokidar@2.1.8: Chokidar 2 will break on node v14+. Upgrade to chokidar 3 with 15x less dependencies.
npm WARN deprecated @hapi/joi@15.1.1: Switch to 'npm install joi'
npm WARN deprecated popper.js@1.16.1: You can find the new Popper v2 at @popperjs/core, this package is dedicated to the legacy v1
added 1314 packages, and audited 1315 packages in 43s
89 packages are looking for funding
  run `npm fund` for details
69 vulnerabilities (2 low, 57 moderate, 10 high)
To address issues that do not require attention, run:
  npm audit fix
To address all issues (including breaking changes), run:
  npm audit fix --force

Run `npm audit` for details.
npm notice 
npm notice New minor version of npm available! 8.1.2 -> 8.3.2
npm notice Changelog: https://github.com/npm/cli/releases/tag/v8.3.2
npm notice Run npm install -g npm@8.3.2 to update!
npm notice 

hekai1229@instance-1:~/local/teslausb/ui$ ./build.sh 
++ dirname ./build.sh
+ BASEDIR=.
+ cd .
+ npm run build

> TeslaUSB@0.1.0 build
> vue-cli-service build


⠧  Building for production...[unplugin-vue-components] component "Index"(/home/hekai1229/local/teslausb/ui/src/components/Viewer/Index.vue) has naming conflicts with other components, ignored.
[unplugin-vue-components] component "Index"(/home/hekai1229/local/teslausb/ui/src/components/Viewer/EpisodeSelect/Index.vue) has naming conflicts with other components, ignored.
⠦  Building for production...

 WARNING  Compiled with 3 warnings                                                                                                     2:26:20 AM

 warning  

asset size limit: The following asset(s) exceed the recommended size limit (244 KiB).
This can impact web performance.
Assets: 
  js/chunk-vendors.71be7ef5.js (586 KiB)

 warning  

entrypoint size limit: The following entrypoint(s) combined asset size exceeds the recommended limit (244 KiB). This can impact web performance.
Entrypoints:
  app (766 KiB)
      css/chunk-vendors.365661c2.css
      js/chunk-vendors.71be7ef5.js
      css/app.4117019d.css
      js/app.55170db3.js


 warning  

webpack performance recommendations: 
You can limit the size of your bundles by using import() or require.ensure to lazy load some parts of your application.
For more info visit https://webpack.js.org/guides/code-splitting/

  File                                   Size                                               Gzipped

  dist/js/chunk-vendors.71be7ef5.js      586.26 KiB                                         190.04 KiB
  dist/js/app.55170db3.js                54.24 KiB                                          16.22 KiB
  dist/css/chunk-vendors.365661c2.css    117.59 KiB                                         15.44 KiB
  dist/css/app.4117019d.css              7.96 KiB                                           1.97 KiB

  Images and other types of assets omitted.
 DONE  Build complete. The dist directory is ready to be deployed.
 INFO  Check out deployment instructions at https://cli.vuejs.org/guide/deployment.html
      
+ mv dist new
+ tar -zcf ../teslausb-ui.tgz -C . '--exclude=*.map' ./new
```
</details>

Can you figure out is there any error during `npm install` and try `npm ci` to make a clean installation of dependcies.

(Fixing dependencies' vulnerabilities needs hard work for deep nested dependencies and possible breaking change. And in this case they can be ignored because we don't put those node services online, our producation is just compiled static js/css which are served by nginx.)

599316527 commented:
After an in-depth investigation, I guess the reason is that you made a force audit fix.

`npm audit fix` tells:

```
ansi-regex  >2.1.1 <5.0.1
Severity: moderate
 Inefficient Regular Expression Complexity in chalk/ansi-regex - https://github.com/advisories/GHSA-93q8-gq69-wqmw
fix available via `npm audit fix --force`
Will install @vue/cli-plugin-babel@3.3.0, which is a breaking change

...
```

In `@vue/cli-plugin-babel@3.3.0`, `preset.js` does not exist.

```
$ cat node_modules/@vue/cli-plugin-babel/package.json | grep version
  "version": "4.5.15",

$  ls node_modules/@vue/cli-plugin-babel          
LICENSE      README.md    codemods     generator.js index.js     logo.png     
migrator     package.json     preset.js
```

```
$cat node_modules/@vue/cli-plugin-babel/package.json | grep version
  "version": "3.3.0",

$ ls node_modules/@vue/cli-plugin-babel
LICENSE      README.md    generator.js index.js     logo.png     package.json
```

marcone reviewed (commented):
> It's possible to push generated files to current repo but I don't think it's a good idea. Repo is not designed to store binary files which cause slow checkout, occupy large amounts of repo size (since all pushed files are in git's history) and may exceeds github's maxiumn limitation sooner or later.

The generated tgz is (currently) 225 KB and expands to 788 KB on the Pi. I don't think that's big enough to worry about exceeding the github size any time soon.

> To clarify releases of OS Images and UIs, publishing them into different releases could be more reasonable? Or UIs can be published to npm's registry which is owned by Github as well.

I'm not sure what you mean by "publishing them to different releases".
My main concern with packaging the UI as a "release" is that [the release page](https://github.com/marcone/teslausb/releases) would then show both flashable images and ui archives, which I think would be confusing (especially since a user would never need to download the ui archive manually). If there is a way to "hide" the ui archives so they don't clutter up the release page, that would be better.
Publishing to npm seems overkill to me, and also somewhat wrong since this is quite specific to teslausb and not generally useful to other projects.

Also, is there a way to have the "node_modules" folder be outside the git project? I prefer to keep build artifacts outside the project folder, but `npm install` puts over 200 MB of stuff in that folder.
And related to that, is the "package-lock.json" file required? It appears to be (re)generated by `npm install`, so does it need to be checked in?

marcone commented:
This was merged with a few tweaks on top

599316527 commented:
> My main concern with packaging the UI as a "release" is that [the release page](https://github.com/marcone/teslausb/releases) would then show both flashable images and ui archives, which I think would be confusing

Fine. I understand your concern. But pushing deployment tgz into repo could be my final option. 

If we seperate UI folder into an individual repo, tgz can be published to it's releases page without any confusion to OS image. How about it?

> Publishing to npm seems overkill to me, and also somewhat wrong since this is quite specific to teslausb and not generally useful to other projects.

Yeah. In this way, we just trust npm as a free object storage. But it just works and no one forbids it.

> Also, is there a way to have the "node_modules" folder be outside the git project?

"node_modules" can be removed once html/js/css are built as well as the entire UI folder is no need to be included in OS Image. Just tgz is used for deployment. 

> is the "package-lock.json" file required?

Yes. It's required by `npm ci` to ensure versions installed while building are same as versions while developing. Using `npm install` may cause versions installed are different among development/deployment environments.

> It appears to be (re)generated by `npm install`

`npm install` creates/updates installed dependencies and their versions into "package-lock.json". When we intend to change  dependencies or versions, we use `npm install` to install and update lock file. Otherwise, we use `npm ci` to ensure dependencies are stable especially on continuous integration environment.

marcone commented:
> > My main concern with packaging the UI as a "release" is that [the release page](https://github.com/marcone/teslausb/releases) would then show both flashable images and ui archives, which I think would be confusing
> 
> Fine. I understand your concern. But pushing deployment tgz into repo could be my final option.
> 
> If we seperate UI folder into an individual repo, tgz can be published to it's releases page without any confusion to OS image. How about it?

It's something I considered. For now I've decided not to go that route, but may reconsider in the future. Might also be an idea to allow specifying the repo to use for the UI, so that alternate UIs can be implemented. That has some possible security implications too though.
 
> > Publishing to npm seems overkill to me, and also somewhat wrong since this is quite specific to teslausb and not generally useful to other projects.
> 
> Yeah. In this way, we just trust npm as a free object storage. But it just works and no one forbids it.
> 
> > Also, is there a way to have the "node_modules" folder be outside the git project?
> 
> "node_modules" can be removed once html/js/css are built as well as the entire UI folder is no need to be included in OS Image. Just tgz is used for deployment.

Right, I was mainly concerned with it being inside the ui folder because for my local development I have setup-teslausb pull the repo files directly from my development machine, so I have to explicitly exclude the node_modules folder. It's not a big deal.
 
> > is the "package-lock.json" file required?
> 
> Yes. It's required by `npm ci` to ensure versions installed while building are same as versions while developing. Using `npm install` may cause versions installed are different among development/deployment environments.
> 
> > It appears to be (re)generated by `npm install`
> 
> `npm install` creates/updates installed dependencies and their versions into "package-lock.json". When we intend to change dependencies or versions, we use `npm install` to install and update lock file. Otherwise, we use `npm ci` to ensure dependencies are stable especially on continuous integration environment.

It seems then that perhaps the README should be saying to use `npm ci`, to ensure that the UI is built with the packages with which it was developed? (right now the README says to use `npm install`, which will update package-lock.json)

599316527 commented:
`npm install` is ok for development because any incompatible issue can be awared during debugging/testing. `npm ci` is recommand for automated environments (eg. Github Actions) in which we hope versions are same as ones in local development environments since packages would be published after being built. If tgz is generated on local machine (rather than github actions) after self-testing and pushed to repo, I think both of them are almost indistinction.

marcone commented:
I tried rebuilding this to fix a bug (it shows a blocking "Failed to fetch video list" dialog on a fresh install), however it appears to no longer build with current or older versions of npm/node (I used to use node 16 to build, but that now results in build errors, and node 17, 18 and 19 all fail in various different ways too)

Any suggestions? (note that I moved this to its own project at https://github.com/marcone/teslausb-webui)

599316527 commented:
I tried to build it and got errors which said missing icons. Is it the same errors you got? After I installed the icon package, build successed. The command will alse add it to package.json.

```
npm i -s veui-theme-dls-icons
```

I think the problem is the newest version of the UI Library([VEUI](https://veui.dev/en-US)) treat icon package as peer dependency so npm won't install it automatically. I recommand using `npm ci` (committing package-lock.json as well) to install dependencies if you are not mean to upgrade their versions. Here is a brief comparison of them. https://www.geeksforgeeks.org/difference-between-npm-i-and-npm-ci-in-node-js/

marcone commented:
> I tried to build it and got errors which said missing icons. Is it the same errors you got?

Yes, that was the error I was getting when building with node 16. It then suggested to run `npm install --save veui-theme-dls-icons/anticlockwise veui-theme-dls-icons/clockwise veui-theme-dls-icons/hamburger veui-theme-dls-icons/pause veui-theme-dls-icons/play veui-theme-dls-icons/undo veui-theme-dls-icons/video` however that failed because it apparently wanted to download from github projects that no longer exist.
With node 18 (the current suggested version on nodejs.org) `npm install` gives an error about node-ipc not being compatible with it.

Using node 16 and installing veui-theme-dls-icons manually fixed the problem. Thanks!
